openstack_controller/tests/test_unit_swift.py

```diff
@@ -0,0 +1,327 @@
+        tags=['keystone_server:http://127.0.0.1:8080/identity'],
+    )
+    args_list = []
+    for call in mock_http_post.call_args_list:
+        args, kwargs = call
+        args_list += list(args)
+    assert args_list.count('http://127.0.0.1:6002/v1') == 0
+    if api_type == ApiType.REST:
+        args_list = []
+        for call in mock_http_post.call_args_list:
+            args, _ = call
+            args_list += list(args)
+        assert args_list.count('http://127.0.0.1:8080/identity/v3/auth/tokens') == 4
+    if api_type == ApiType.SDK:
+        assert session_auth.get_access.call_count == 4
+    assert '`swift` component not found in catalog' in caplog.text
+
+
+@pytest.mark.parametrize(
+    ('mock_http_get', 'instance'),
+    [
+        pytest.param(
+            {'http_error': {'/v1/AUTH_1e6e233e637d4d55a50a62b63398ad15': MockResponse(status_code=500)}},
+            configs.REST,
+            id='api rest',
+        ),
+        pytest.param(
+            {'http_error': {'/v1/AUTH_1e6e233e637d4d55a50a62b63398ad15': MockResponse(status_code=500)}},
+            configs.SDK,
+            id='api sdk',
+        ),
+    ],
+    indirect=['mock_http_get'],
+)
+@pytest.mark.usefixtures('mock_http_get', 'mock_http_post', 'openstack_connection')
+def test_response_time_exception(aggregator, check, dd_run_check, mock_http_get):
+    dd_run_check(check)
+    aggregator.assert_metric(
+        'openstack.swift.response_time',
+        count=0,
+    )
+    aggregator.assert_service_check(
+        'openstack.swift.api.up',
+        status=AgentCheck.CRITICAL,
+        tags=['keystone_server:http://127.0.0.1:8080/identity'],
+    )
+    args_list = []
+    for call in mock_http_get.call_args_list:
+        args, kwargs = call
+        args_list += list(args)
+    assert args_list.count('http://127.0.0.1:6002/v1') == 0
+
+
+@pytest.mark.parametrize(
+    ('instance', 'api_type'),
+    [
+        pytest.param(
+            configs.REST,
+            ApiType.REST,
+            id='api rest',
+        ),
+        pytest.param(
+            configs.SDK,
+            ApiType.SDK,
+            id='api sdk',
+        ),
+    ],
+)
+@pytest.mark.usefixtures('mock_http_get', 'mock_http_post', 'openstack_connection')
+def test_response_time(aggregator, check, dd_run_check, mock_http_get, api_type):
+    dd_run_check(check)
+    aggregator.assert_metric(
+        'openstack.swift.response_time',
+        count=1,
+        tags=['keystone_server:http://127.0.0.1:8080/identity'],
+    )
+    aggregator.assert_service_check(
+        'openstack.swift.api.up',
+        status=AgentCheck.OK,
+        tags=['keystone_server:http://127.0.0.1:8080/identity'],
+    )
+    args_list = []
+    for call in mock_http_get.call_args_list:
+        args, kwargs = call
+        args_list += list(args)
+    if api_type == ApiType.REST:
+        assert args_list.count('http://127.0.0.1:6002/v1/AUTH_1e6e233e637d4d55a50a62b63398ad15') == 2
```

Comment:
Looking at the `kwargs` for each call in mock_http_get.call_args_lists, the first call lacks the following argument in kwargs that the second call has: `'params': {'format': 'json'}`. This is because Swift is a special case where we only add the format=json parameter in getting all of the non-response-time metrics. As such, there are two calls--one for _report_response_time (without the parameter specified), and the second to _report_containers (with the parameter specified).

The reason why this logic only applies to the rest api and not the sdk api is that we add {'format': 'json'} in params for _report_containers in the rest api, while on the other hand, format=json is explicitly add to the url used to make the request.

Reply:
Why can we not first call with the 'params' so we avoid calling the endpoint twice?

Reply:
Manually including `format=json` as a parameter in the rest api for the first call just causes the identical call to appear twice:
`('http://127.0.0.1:6002/v1/AUTH_1e6e233e637d4d55a50a62b63398ad15',)
{'params': {'format': 'json'}, 'auth': ('admin', 'password'), 'cert': None, 'headers': OrderedDict([('User-Agent', 'Datadog Agent/0.0.0'), ('Accept', '*/*'), ('Accept-Encoding', 'gzip, deflate'), ('X-Auth-Type', '6e39099cccde4f809b003d9e0dd09304'), ('X-Auth-Token', 'token_6e39099cccde4f809b003d9e0dd09304')]), 'proxies': None, 'timeout': (10.0, 10.0), 'verify': True, 'allow_redirects': True}
('http://127.0.0.1:6002/v1/AUTH_1e6e233e637d4d55a50a62b63398ad15',)
{'params': {'format': 'json'}, 'auth': ('admin', 'password'), 'cert': None, 'headers': OrderedDict([('User-Agent', 'Datadog Agent/0.0.0'), ('Accept', '*/*'), ('Accept-Encoding', 'gzip, deflate'), ('X-Auth-Type', '6e39099cccde4f809b003d9e0dd09304'), ('X-Auth-Token', 'token_6e39099cccde4f809b003d9e0dd09304')]), 'proxies': None, 'timeout': (10.0, 10.0), 'verify': True, 'allow_redirects': True}`

Typically, when there is no parameter specified, only one call appears (here is the example for heat):
`('http://127.0.0.1:8004/heat-api',)
{'auth': None, 'cert': None, 'headers': OrderedDict([('User-Agent', 'Datadog Agent/0.0.0'), ('Accept', '*/*'), ('Accept-Encoding', 'gzip, deflate'), ('X-Auth-Token', <MagicMock name='mock.session.auth.get_token()' id='4491112528'>)]), 'proxies': None, 'timeout': (10.0, 10.0), 'verify': True, 'allow_redirects': True}`